Chore: Allow MenuItems to have `role=menuitem` when URL is set

tomratcliffe opened the pull request:
**What is this feature?**
Allows MenuItems to have role=menuitem when a URL is set

**Why do we need this feature?**

I ran into this when trying to write a test that checks the correct menu items are present in a dropdown, based on permissions. It took a little unravelling to work out why some of the expected elements were missing from my test

**Who is this feature for?**

UI devs + all users. More consistent screen reader experience? Up for discussion as it's gone from rendering as a "link" to a "menu item" - but that's accurate?

tomratcliffe commented:
Up for discussion as this does change elements from `link` to `menuitem` 🤷 Not sure what's more accurate, thoughts? I am happy to ditch this change altogether and tweak my original test elsewhere, it was just confusing that I didn't get all of the menuitems I expected when querying the dropdown

ckbedwell commented:
I'm on the fence whether I agree with this change or not. The two places which I'm inspecting in particular for this are:

![](https://github.com/grafana/grafana/assets/6906380/ba8e688e-6425-447a-93c9-6726f2a7afab)
*The Grafana help 'menu', which lives in the top bar of Grafana.*

![](https://github.com/grafana/grafana/assets/6906380/9831cef4-3f1d-4ae3-bc43-1f75797676d0)
*The Grafana panel options menu, which is on every dashboard panel.*

Despite they share the same design pattern I'd argue the first one is a disclosure widget with a set of extra options (that mostly happen to be links except one) and _not_ a menu.

Where the second one is a _true_ menu in the sense of what you might find in a native application.

So, in one sense I agree with this change in a theoretical way, but in a pragmatic way of how the `<Menu />` and <`MenuItem />`s are currently implemented across Grafana I think this makes accessibility worse.

Using my screen reader I am _very_ clear on the behaviour of each item in the help 'menu' and know what change of context to expect if I were to interact with each item (it is especially helpful to know which are links because I can then choose to open them in a new tab if I wish), with this new change I don't really know what is going to happen when I select them as all the other menuitems on the page cause some sort of in-app context change.

For reference, I am using:
[Don't use Aria Menu Roles for site Nav](https://adrianroselli.com/2017/10/dont-use-aria-menu-roles-for-site-nav.html) by Adrian Roselli
[Menus & Menu Buttoons](https://inclusive-components.design/menus-menu-buttons/) by Inclusive Components

And despite I don't necessarily agree with this implementation, I am also looking at:
[Navigation Menubar Example
](https://www.w3.org/WAI/ARIA/apg/patterns/menubar/examples/menubar-navigation/) by the APG, but they do link to [Example Disclosure Navigation Menu](https://www.w3.org/WAI/ARIA/apg/patterns/disclosure/examples/disclosure-navigation/) which I do agree with.

tomratcliffe committed:
Allow passing a role into a MenuItem component, and fallback to default behaviour if not passed

tomratcliffe commented:
I've updated the behaviour now - I removed the default role, and fallback to `menuitem` within the component.

If there's no URL, uses either the passed role, or falls back to `menuitem`.
If there is a URL, will use the passed role, or fallback to undefined, which preserves `role = link`

ckbedwell reviewed (approved):
Awesome! LGTM 🚀

gillesdemey commented on packages/grafana-ui/src/components/Menu/MenuItem.tsx:
```diff
@@ -29,7 +37,7 @@ export interface MenuItemProps<T = unknown> {
   icon?: IconName;
   /** Role of the menu item */
-  role?: string;
+  role?: AriaRole;
```
👍
